### PR TITLE
Implement local AIFS trigger in amd-ras service

### PR DESCRIPTION
**Pull request by abinayaddhandapani**
Added the new RAS configuration parameters AifsArmed and AifsSigId.

If AifsArmed is true and valid mca bank is > 0 , Check if current-failure SigID matches any of the AifsSigId entries in the AifsSignatureIdList. If SigID matches, start the AIFS flow.

Signed-off-by: Abinaya Dhandapani <adhandap@amd.com>

### CODE REVIEW

File: src/fatal_error.cpp [hunk 1]
```diff
@@ -1046,8 +1222,8 @@ void P1PmicGlEventHandler()
 
     if (gpioLineEvent.event_type == gpiod::line_event::RISING_EDGE)
     {
-        std::string ras_err_msg =
-            "P1 DIMM G-L PMIC FATAL Error detected. System will be power off";
+        std::string ras_err_msg = "P1 DIMM G-L PMIC FATAL Error detected. "
+                                  "System will be power off";
```
Comment (mahkurap) [inline src/fatal_error.cpp:1223]: Change to "System will be powered off"

File: src/fatal_error.cpp [hunk 2]
```diff
@@ -956,8 +1132,8 @@ void P0PmicAfEventHandler()
 
     if (gpioLineEvent.event_type == gpiod::line_event::RISING_EDGE)
     {
-        std::string ras_err_msg =
-            "P0 DIMM A-F PMIC FATAL Error detected. System will be power off";
+        std::string ras_err_msg = "P0 DIMM A-F PMIC FATAL Error detected. "
+                                  "System will be power off";
```
Comment (mahkurap) [inline src/fatal_error.cpp:1133]: change "power" to "powered"

Comment (mahkurap) [reply src/fatal_error.cpp:1133]: if there is a chance of dependent client applications on this message, don't change it.

File: src/fatal_error.cpp [hunk 3]
```diff
@@ -1016,8 +1192,8 @@ void P1PmicAfEventHandler()
 
     if (gpioLineEvent.event_type == gpiod::line_event::RISING_EDGE)
     {
-        std::string ras_err_msg =
-            "P1 DIMM A-F PMIC FATAL Error detected. System will be power off";
+        std::string ras_err_msg = "P1 DIMM A-F PMIC FATAL Error detected. "
```
Comment (mahkurap) [inline src/fatal_error.cpp:1192]: May be this is a late in the game comment.  To the event handler pass the AlertLine information as input parameter, then we could avoid duplicating the code. 
void PmicAfEventHandler(*pmicAlertLine)

Comment (mahkurap) [reply src/fatal_error.cpp:1192]: How are we handling the systems with one SoC?

File: inc/Config.hpp [hunk 4]
```diff
@@ -55,6 +57,9 @@ class Configuration
     static void setPcieAerThresholdEn(bool);
     static bool getPcieAerThresholdEn();
 
+    static void setAifsArmed(bool);
+    static bool getAifsArmed();
+
```
Comment (mahkurap) [inline inc/Config.hpp:62]: nit: extra new line

File: inc/ras.hpp [hunk 5]
```diff
@@ -79,6 +80,7 @@ extern "C" {
 #define RUNTIME_PCIE_ERR ("RUNTIME_PCIE_ERROR")
 #define RUNTIME_DRAM_ERR ("RUNTIME_DRAM_ERROR")
 #define FATAL_ERR ("FATAL")
+#define EVENT_SUBSCRIPTION_FILE ("/var/lib/bmcweb/eventservice_config.json")
```
Comment (mahkurap) [inline inc/ras.hpp:83]: EVENT_SUBSCRIPTION_FILE defined but never used.

File: src/dbus_ras.cpp [hunk 6]
```diff
@@ -308,6 +308,27 @@ void CreateDbusInterface()
             return 1;
         });
 
+    bool AifsArmed = Configuration::getAifsArmed();
+    configIface->register_property("AifsArmed", AifsArmed,
+                                   [](const bool& requested, bool& resp) {
+                                       resp = requested;
+                                       Configuration::setAifsArmed(resp);
+                                       updateConfigFile("AifsArmed", resp);
+                                       return 1;
```
Comment (mahkurap) [inline src/dbus_ras.cpp:317]: magic number

File: src/fatal_error.cpp (hunk 3 above)
Comment (abinayaddhandapani) [reply src/fatal_error.cpp:1192]: PMIC DIMM Error changes is implemented by Mohsen.
Hi @modolaty , Can you please help to answer Mahesh review comments ?

File: inc/ras.hpp (hunk 5 above)
Comment (abinayaddhandapani) [reply inc/ras.hpp:83]: This is used in fatal_error.cpp file -> API harvest_ras_errors()

File: src/main.cpp [hunk 7]
```diff
@@ -599,112 +620,163 @@ void performPlatformInitialization()
...
+            if (*currentTimerEnable == false)
+            {
+
+                sdbusplus::bus::bus bus = sdbusplus::bus::new_default();
+                std::string CurrentTimerUse = getProperty<std::string>(
+                    bus, "xyz.openbmc_project.Watchdog",
+                    "/xyz/openbmc_project/watchdog/host0",
+                    "xyz.openbmc_project.State.Watchdog", "CurrentTimerUse");
+
+                if (CurrentTimerUse ==
```
Comment (ojayanth) [inline src/main.cpp:760]: Is this flow is good from BMC runtime reboot use case?

File: src/main.cpp [hunk 8]
```diff
@@ -591,6 +594,24 @@ static void currentHostStateMonitor()
                 std::ofstream outFile(dramCeccErrorFile.data());
                 outFile << std::setw(INDEX_4) << j << std::endl;
             }
+
+            if (*currentHostState !=
```
[the remark this answers is not in the document]
Comment (ojayanth) [reply src/main.cpp:598]: Looks like openbmc integ branch doesn't use PSM for state management, and also not seeing these sate usage in amd-power-control . looks like you can ignore my comments. we can relook this logic in ing-sp7 development

File: inc/Config.hpp (hunk 4 above)
Comment (abinayaddhandapani) [reply inc/Config.hpp:62]: The extra new line is added between every set and get API for improved readability and clarity.

File: src/dbus_ras.cpp [hunk 9]
```diff
@@ -308,6 +308,27 @@ void CreateDbusInterface()
...
+    std::vector<std::pair<std::string, std::string>> AifsSignatureId =
+        Configuration::getAllAifsSignatureId();
+    configIface->register_property(
+        "AifsSignatureId", AifsSignatureId,
+        [](const std::vector<std::pair<std::string, std::string>>& requested,
+           std::vector<std::pair<std::string, std::string>>& resp) {
+            resp = requested;
+            Configuration::setAllAifsSignatureId(resp);
+            updateConfigFile("AifsSignatureId", resp);
+            return 1;
```
[the remark this answers is not in the document]
Comment (supven01) [reply src/dbus_ras.cpp:329]: +1

File: src/fatal_error.cpp (hunk 1 above)
Comment (supven01) [reply src/fatal_error.cpp:1223]: +1

File: src/fatal_error.cpp (hunk 2 above)
Comment (supven01) [reply src/fatal_error.cpp:1133]: Dependent client applications should not be using string message. If they do, we need to ensure that they are not doing it.

File: src/dbus_ras.cpp (hunk 6 above)
Comment (supven01) [reply src/dbus_ras.cpp:317]: +1

File: src/main.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (supven01) [reply src/main.cpp:596]: Can you elaborate more? We have Host states "on" and "off" defined in amd-power-control based off of the POWER_GOOD signal and that's working as expected.

File: src/dbus_ras.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (abinayaddhandapani) [reply src/dbus_ras.cpp]: Fixed

File: src/fatal_error.cpp (hunk 1 above)
Comment (abinayaddhandapani) [reply src/fatal_error.cpp]: Fixed

File: src/dbus_ras.cpp (hunk 6 above)
Comment (abinayaddhandapani) [reply src/dbus_ras.cpp]: Fixed

File: src/fatal_error.cpp [hunk 10]
```diff
@@ -708,6 +708,106 @@ void harvest_fatal_errors(uint8_t info, uint16_t numbanks, uint16_t bytespermca)
     }
 }
 
+std::vector<uint32_t> hexstring_to_vector(const std::string& hexString)
+{
+    std::vector<uint32_t> result;
+    size_t len = hexString.length();
+
+    // Skip the "0x" prefix if present
+    size_t start = (hexString.substr(0, 2) == "0x") ? 2 : 0;
```
Comment (supven01) [inline src/fatal_error.cpp:717]: nit: magic numbers 2

File: src/fatal_error.cpp [hunk 11]
```diff
@@ -708,6 +708,106 @@ void harvest_fatal_errors(uint8_t info, uint16_t numbanks, uint16_t bytespermca)
...
+
+    // Process the string in chunks of 8 characters (32 bits)
```
Comment (supven01) [inline src/fatal_error.cpp:719]: What's the reason behind processing 8 characters at a time?

**Pull request merged**